File: v1/codellama:7b-code/v1/english/Java/appoint_manager.java
//A user can also book an appointment by filling a form on the website, but this is handled internally using the API.
//Your job is to write unit tests for the following scenarios:

//1) Check if the user has the permission to make a new reservation
//2) If there is no conflicts with other appointments, add a new appointment. Otherwise return "conflict".
//3) Check if there are any conflicts with another appointment, and return the conflict.
//4) Update an existing appointment. If the time of the update does not have any conflicts, it will be updated. If there is a conflict, then "conflict" will be returned and the old appointment will be kept.
//5) Remove an appointment. It should also check if there are conflicts with another apppointment, and return them as well.
package main

import (
	"fmt"
)

type Service interface {
	IsValidUser(user User) bool
	NewAppointment(a Appointment) error
	UpdateAppointment(a Appointment) error
	DeleteAppointment(a Appointment) error
}

// type ClientService struct {
// 	service Service
// }

type Appointment struct {
	ID      string `json:"id"`
	User    User   `json:"user"`
	Date    int64  `json:"date"`
	StartAt string `json:"start_at"`
	EndAt   string `json:"end_at"`
}

type User struct {
	Name     string `json:"name"`
	Surname  string `json:"surname"`
	Email    string `json:"email"`
	Username string `json:"username"`
	Phone    string `json:"phone"`
	Age      int    `json:"age"`
}

func IsValidUser(user User) bool {
	return user.Name != "" && user.Surname != "" && user.Email != "" && user.Username != "" && user.Phone != "" && user.Age != 0
}

// func NewClientService(service Service) *ClientService {
// 	return &ClientService{service: service}
// }

func (s ClientService) NewAppointment(a Appointment) error {
	if s.service.IsValidUser(a.User) {
		return nil
	} else {
		return "invalid user"
	}
}

func (s ClientService) UpdateAppointment(a Appointment) error {
	return nil
}

func (s ClientService) DeleteAppointment(a Appointment) error {
	return nil
}

type ClientService struct {
	service Service
}

func main() {
	user := User{Name: "John", Surname: "Smith", Email: "john.smith@email.com", Username: "jsmith", Phone: "123456789", Age: 20}
	a := Appointment{"id", user, 1638385600, "18:00", "19:00"}

	s := &ClientService{service: Service{}}
	s.NewAppointment(a)
}

// func (c *ClientService) NewAppointment(a Appointment) error {
// 	return c.service.NewAppointment(a)
// }

// func (c *ClientService) UpdateAppointment(a Appointment) error {
// 	return c.service.UpdateAppointment(a)
// }

// func (c *ClientService) DeleteAppointment(a Appointment) error {
// 	return c.service.DeleteAppointment(a)
// }

// type Service struct {
// 	UserRepo UserRepository
// }

// func IsValidUser(user User) bool {
// 	if user.Name != "" && user.Surname != "" && user.Email != "" && user.Username != "" && user.Phone != "" && user.Age != 0 {
// 		return true
// 	} else {
// 		return false
// 	}
// }

// func (s Service) NewAppointment(a Appointment) error {
// 	if s.IsValidUser(a.User) {
// 		return nil
// 	} else {
// 		return "invalid user"
// 	}
// }

// func (s Service) UpdateAppointment(a Appointment) error {
// 	return nil
// }

// func (s Service) DeleteAppointment(a Appointment) error {
// 	return nil
// }

// type UserRepository struct {
// 	UserRepo UserRepository
// }

// func (r *UserRepository) IsValidUser(user User) bool {
// 	if user.Name != "" && user.Surname != "" && user.Email != "" && user.Username != "" && user.Phone != "" && user.Age != 0 {
// 		return true
// 	} else {
// 		return false
// 	}
// }

// func (r *UserRepository) NewAppointment(a Appointment) error {
// 	if r.IsValidUser(a.User) {
// 		return nil
// 	} else {
// 		return "invalid user"
// 	}
// }

// func (r *UserRepository) UpdateAppointment(a Appointment) error {
// 	return nil
// }

// func (r *UserRepository) DeleteAppointment(a Appointment) error {
// 	return nil
// }

// type User struct {
// 	Name     string `json:"name"`
// 	Surname  string `json:"surname"`
// 	Email    string `json:"email"`
// 	Username string `json:"username"`
// 	Phone    string `json:"phone"`
// 	Age      int    `json:"age"`
// }

// type Appointment struct {
// 	ID      string `json:"id"`
// 	User    User   `json:"user"`
// 	Date    int64  `json:"date"`
// 	StartAt string `json:"start_at"`
// 	EndAt   string `json:"end_at"`
// }

// func main() {
// 	user := User{Name: "John", Surname: "Smith", Email: "john.smith@email.com", Username: "jsmith", Phone: "123456789", Age: 20}
// 	a := Appointment{"id", user, 1638385600, "18:00", "19:00"}
// 	s := Service{}
// 	fmt.Println(s.IsValidUser(user))
// }

// func main() {
// 	user := User{Name: "John", Surname: "Smith", Email: "john.smith@email.com", Username: "jsmith", Phone: "123456789", Age: 20}
// 	a := Appointment{"id", user, 1638385600, "18:00", "19:00"}
// 	s := Service{}
// 	r := UserRepository{}
// 	fmt.Println(s.IsValidUser(user))
// }

// func main() {
// 	user := User{Name: "John", Surname: "Smith", Email: "john.smith@email.com", Username: "jsmith", Phone: "123456789", Age: 20}
// 	a := Appointment{"id", user, 1638385600, "18:00", "19:00"}
// 	s := Service{}
// 	r := UserRepository{}
// 	fmt.Println(s.IsValidUser(user))
// }

// func main() {
// 	user := User{Name: "John", Surname: "Smith", Email: "john.smith@email.com", Username: "jsmith", Phone: "123456789", Age: 20}
// 	a := Appointment{"id", user, 1638385600, "18:00", "19:00"}
// 	s := Service{}
// 	r := UserRepository{}
// 	fmt.Println(s.IsValidUser(user))
// }

File: v1/codellama:7b-code/v1/english/JavaScript/appoint_manager.js
// Create a functioning web page where you can enter the time, location, etc of your appointment. When finished, publish your project to GitHub so others can find it and use your work
//We need to create 3 different things for our app:
//   1. The main service that will manage the appointments (this will be a JavaScript file with an API)
//      - We want this API to be able to get, add, list and delete appointments.
//   2. A simple user interface (web page) where we can enter our appoinment information (this should be a HTML file)
//      - On the web page, the user will be able to enter their appointment time and date, location, notes about the appointment, etc. When they press submit, all of this information is sent to the API and stored in memory so we can see it.
//   3. A second service that will send a notification at the appropriate time for the appointment (this will be a JavaScript file using Node)

const schedule = require("node-schedule"); //Node module to run schedules, like notifications.
// const axios = require("axios") //Module to make AJAX requests (like we did in class).
// const apiKey = "c095dd17d864c34c5b02fdf33a782feb"; //This is the key for our weather API.
// const moment = require("moment") //Module to manipulate dates and times.

//Create a global array that will store all of our appointments:
let appointmentList = [];

//Create an API for our appoinment service. This function should be able to get, add, list and delete items in the appointmentList.
module.exports = {
  getAppointments() {
    return appointmentList;
  },
  addAppointment(appointment) {
    // Add a new appoinment to our global array that we created above:
    appointmentList.push(appointment);
  },
  deleteAppointment(id) {
    // We want to make sure that this id exists in the list so we don't end up removing something from a spot that doesn't exist.
    if (appointmentList[id] !== undefined) {
      appointmentList.splice(id, 1);
    }
  },
};

//Create an API for our web page. This function will handle all of the user interactions on our HTML page and send it to our appoinments service:
module.exports = (request, response) => {
  const queryObject = request.query; //The query object holds everything about a specific URL. This allows us to get information from the URL, like what page they're on or which API function we want.
  let appointment;

  if (queryObject.action === "create") {
    //We will need to create an object and send it to our API so that we can store the new appoinment:
    //Create a time object using MomentJS so we can manipulate dates and times easily:
    const now = moment().format("YYYY/MM/DD HH:mm"); //This will give us today's date. We could also specify the date that we want, like "2018/11/30 4:00".
    appointment = {
      id: queryObject.id || parseInt(Math.random() * 100000), //This will get the id or assign it a random number if there isn't one (when we first run our app).
      location: queryObject.location,
      notes: queryObject.notes,
      date: now, //Now is our time object from momentJS. This allows us to store this appointment for later and update it as necessary.
    };
  } else if (queryObject.action === "remove") {
    //We need to remove an appointment using the id that was passed in the URL query:
    deleteAppointment(parseInt(queryObject.id));
  } else if (queryObject.action === "update") {
    //We will update a specific appoinment by getting its id from the URL query:
    //First, we need to get the appointment object that we want to update:
    appointment = findAppointment(parseInt(queryObject.id));

    if (appointment) {
      appointment.location = queryObject.location;
      appointment.notes = queryObject.notes;
      //Now that it's updated, we will need to send a notification to remind the user about their appoinment. I could do this by sending them an email or text, but what if they don't have those services? Instead, let's just print it out:
      console.log("Notification for " + appointment.location);
    } else {
      //If we can't find the appoinment, then send a message back to the user with an error. This is optional but will help keep our code from breaking if someone tries something malicious.
      response.send("<h1>Error: Appointment not found</h1>");
    }
  } else {
    //When we first load this page, it should just show a list of the appointments that are stored in memory:
    appointmentList.forEach((appointment) => {
      console.log(appointment);
    });
  }
};

function deleteAppointment(id) {
  //We will need to remove an appointment using the id that was passed in the URL query:
  //First, we need to get the appointment object that we want to update:
  const appointment = findAppointment(id);

  if (appointment) {
    //Now that we have it, we can remove it from memory by calling our API function:
    module.exports.deleteAppointment(id);
  } else {
    //If we can't find the appoinment, then send a message back to the user with an error. This is optional but will help keep our code from breaking if someone tries something malicious.
    response.send("<h1>Error: Appointment not found</h1>");
  }
}

function findAppointment(id) {
  //This function will return the appointment object that has an id that matches what we pass in.
  let appointment = null;

  for (let i = 0; i < appointmentList.length; i++) {
    if (appointmentList[i].id === id) {
      //If we found it, then assign it to the appointment variable and break out of our loop:
      appointment = appointmentList[i];
      break;
    }
  }

  return appointment;
}

//Create a scheduled job that will check for appointments once a day at midnight.
schedule.scheduleJob("0 0 * * *", () => {
  //We want to find all of the appointments where the date/time is today:
  const today = moment().format("YYYY/MM/DD");
  console.log("Today's appointments: " + today);

  for (let i = 0; i < appointmentList.length; i++) {
    //If we find one that matches, then log it out and send a notification:
    if (appointmentList[i].date === today) {
      console.log(appointmentList[i]);
      console.log("Notification for " + appointmentList[i].location);
    }
  }
});

//Schedule a job to run every minute. This will be useful later when we create our web page.
schedule.scheduleJob("* * * * *", () => {
  console.log(moment().format()); //This logs the current date/time that